routes/api.js

```diff
@@ -145,36 +139,56 @@ router.post('/kernels', bodyParser.json({ type: 'text/plain' }), function(req, r
-            error('Error proxying kernel creation request:' + err.toString());
-            return res.status(500).end();
-        }
-        // Store the notebook path for use within the WS proxy.
-        var notebookPathHeader = req.headers['x-jupyter-notebook-path'];
-        var sessionId = req.headers['x-jupyter-session-id'];
-        if (!notebookPathHeader || !sessionId) {
-            error('Missing notebook path or session ID headers');
-            return res.status(500).end();
-        }
-        var matches = notebookPathHeader.match(/^\/(?:dashboards(-plain)?)?(.*)$/);
-        if (!matches) {
-            error('Invalid notebook path header');
-            return res.status(500).end();
-        }
-        // Store notebook path for later use
-        sessions[sessionId] = matches[2];
+    // Store the notebook path for use within the WS proxy.
+    var notebookPathHeader = req.headers['x-jupyter-notebook-path'];
+    var sessionId = req.headers['x-jupyter-session-id'];
+    if (!notebookPathHeader || !sessionId) {
+        error('Missing notebook path or session ID headers');
+        return res.status(500).end();
+    }
 
-        // Pass the kernel gateway response back to the client.
-        res.set(response.headers);
-        res.status(response.statusCode).json(body);
-    });
+    var matches = notebookPathHeader.match(/^\/(?:dashboards(-plain)?)?(.*)$/);
+    if (!matches) {
+        error('Invalid notebook path header');
+        return res.status(500).end();
+    }else{
+        var notebookPath = matches[2];
+        // Store notebook path for later use
+        sessions[sessionId] = notebookPath;
+
+        nbstore.get(notebookPath)
+        .then(function success(notebook){
+          debug(notebook.metadata);
+          if (notebook.metadata.kernelspec.name){
+              var kernelName = notebook.metadata.kernelspec.name;
+              debug("Notebook kernel name found: " + kernelName);
+              if (kernelName === "apache_toree"){
+                  kernelName = "scala";
+              }
+              req.body.name = kernelName;
+          }else{
+              // Default to Python 3
+              debug("Notebook kernel name not found, defaulting to Python 3");
+              req.body.name = "python3";
+          }
+          console.log(req.body);
+          // Pass the (modified) request to the kernel gateway.
+          request({
+            url: urljoin(kgUrl, kgBaseUrl, '/api/kernels'),
+            method: 'POST',
+            headers: headers,
+            json: req.body
+          }, function(err, response, body) {
+            if(err) {
+                error('Error proxying kernel creation request:' + err.toString());
+                return res.status(500).end();
+            }
+            // Pass the kernel gateway response back to the client.
+            res.set(response.headers);
+            res.status(response.statusCode).json(body);
+          });
+        });
+      }
```

Comment:
I know this is WIP, so I won't have full review comments, but I just want to say that you should adhere to the coding style already in the file:
1. Indentation is 4 spaces (not 2).
2. Add appropriate spaces around keywords (i.e. `} else {`).